src/components/Tokens/TokenDetails/index.tsx

```diff
@@ -64,6 +64,28 @@ function useOnChainToken(address: string | undefined, skip: boolean) {
   }
 }
 
+// Selects most relevant token based on data available, preferring native > query > on-chain
+// Token will be null if still loading from on-chain, and undefined if unavailable
+function useRelevantToken(
+  address: string | undefined,
+  pageChainId: number,
+  tokenQueryData: TokenQueryData | undefined
+) {
+  const { chainId: activeChainId } = useWeb3React()
+  const queryToken = useMemo(() => {
+    if (!address) return undefined
+    if (address === NATIVE_CHAIN_ID) return nativeOnChain(pageChainId)
+    if (tokenQueryData) return new QueryToken(tokenQueryData)
+    return undefined
+  }, [pageChainId, address, tokenQueryData])
+  // fetches on-chain token if query data is missing and page chain matches global chain (else fetch won't work)
+  const skipOnChainFetch = Boolean(queryToken) || pageChainId !== activeChainId
+  const onChainToken = useOnChainToken(address, skipOnChainFetch)
+  const token = useMemo(() => queryToken ?? onChainToken, [queryToken, onChainToken])
+
+  return { token, didFetchFromChain: !queryToken }
```

Comment:
This must be memoized, as it will create a new object with every call.

    return useMemo(() => ({
      token: queryToken ?? onChainToken,
      didFetchFromChain: !queryToken,
    }, [queryToken, onChainToken])